### ISSUE TITLE
[aws-certificatemanager] Create certificate in us-east-1 and use it in a different region

### ISSUE THREAD

**Issue by dbartholomae** [aws/aws-cdk#9274]
### The Question

I'm currently setting up AWS Cognito with a custom domain via AWS CDK. Our stack lives in eu-central-1, but as I understand the certificate for the custom domain has to live in us-east-1. How can I share the certificate to the AWS Cognito setup?

### Environment

  - **CDK CLI Version:** 1.54.0
  - **Module Version:** 1.54.0
  - **Node.js Version:** v12.16.0
  - **OS:** Windows 10
  - **Language (Version):** TypeScript (3.9.7)

### Other information
There's a [discussion around this](https://github.com/aws/aws-cdk/issues/8232) already which indicates there might not be a solution to this?

**Comment by njlynch** [comment]
Hi @dbartholomae,

You generally have two different options for using a certificate cross-region.

1. Create a dedicated stack for the certificate in us-east-1, and import the certificate in your Cognito stack (in eu-central-1).

2. Use the [DnsValidatedCertificate](https://github.com/aws/aws-cdk/blob/master/packages/%40aws-cdk/aws-certificatemanager/lib/dns-validated-certificate.ts) construct in your Cognito stack, which is a custom resource that can request a certificate cross-region.

Let me know if that helps!

**Comment by dbartholomae** [comment]
Thanks for the answer! It is good to know that DnsValidatedCertificate is able to request cross-region. Unfortunately the domain in question is not managed by Route53.
For 1: How would I do this import in CDK, given that both stacks are defined in CDK? Do I need to change the CI setup to run multiple times for multiple stacks and somehow pass around the result from one call to the other?

**Comment by njlynch** [comment]
If you already have a Certificate created and defined, you can skip the second stack and just import it directly in your Cognito stack.

```ts
const certificate = acm.Certificate.fromCertificateArn(this, 'MyImportedCert', 'arn:aws:acm:us-east-1:123456789012:certificate/12345678-1234-1234-1234-123...');
```

If you are creating the certificate in one stack (and validating via email or manually with another DNS provider), then you can pass a reference from one stack to another. Something like this:

```ts
class MyCertStack extends cdk.Stack {
  public readonly certificate: acm.Certificate;

  constructor(scope: cdk.App, id: string, props?: cdk.StackProps) {
    super(scope, id, props);
    certificate = new acm.Certificate(...);
  }
}

interface MyCognitoStackProps extends cdk.StackProps {
  certificate: acm.Certificate;
}

class MyCognitoStack extends cdk.Stack {
  constructor(scope: cdk.App, id: string, props: MyCognitoStackProps) {
    const certificate = props.certificate;
    // Define your Cognito pool here, using the above certificate
  }
}

const certStack = new MyCertificateStack(app, 'MyCerts');
new MyCognitoStack(app, 'MyCognito', { certificate: certStack.certificate });
```

**Comment by dbartholomae** [comment]
I'm trying the second solution, but so far I always get a cross-stack reference error because the certificate is in us-east-1, but the other stack in eu-central-1.

**Comment by njlynch** [comment]
Ah, that's correct. Sorry for the misinformation earlier.

Unfortunately, we currently only support cross-environment (region/account) references in instances where we can assign the physical names to the resources; in cases like certificates, where there is only the ARN, we don't currently have a solution.

See https://github.com/aws/aws-cdk/issues/8232#issuecomment-635639322 for one work-around suggestion of creating a custom resource to do the heavy lifting for you. It looks like there are various community-owned solutions to this; I found https://www.npmjs.com/package/@henrist/cdk-cross-region-params as one example. I haven't tested this and can't support it, but it looks like it could be used to pass the certificate ARN cross-stack. YMMV. :)

**Comment by dbartholomae** [comment]
Thanks! Using SSM is actually quite a smart solution. I'll look deeper into it.

**Comment by robertofd1995** [comment]
Did aws cdk added officially support for this? I am also struggling with the same problem.

**Comment by lenfree** [comment]
@robertd not sure if are still stuck with this. If you still are, I think you might be looking  for [create cross region AMC](https://docs.aws.amazon.com/cdk/api/latest/docs/@aws-cdk_aws-certificatemanager.DnsValidatedCertificate.html#class-dnsvalidatedcertificate-construct).

**Comment by A-ndy-git** [comment]
+1 for this feature. I know there are some suggested workaround above, but would be nice to see an easier method to implement this.

My scenario is an Edge API in `eu-west-2` with a custom domain. The certificate for the custom domain has to be created in `us-east-1`. Attempting to import the certificate via ARN outputs the `cannot find ARN in eu-west-2` error.

**Comment by Booligoosh** [comment]
Thank you @lenfree, you just saved me hours of headaches 🤗

**Comment by jedrekdomanski** [comment]
> @robertofd1995 not sure if are still stuck with this. If you still are, I think you might be looking for [create cross region ACM](https://docs.aws.amazon.com/cdk/api/latest/docs/aws-certificatemanager-readme.html#cross-region-certificates).

You saved my day ❤️

**Comment by Jaftem** [comment]
> @robertofd1995 not sure if are still stuck with this. If you still are, I think you might be looking for [create cross region ACM](https://docs.aws.amazon.com/cdk/api/latest/docs/aws-certificatemanager-readme.html#cross-region-certificates).

Using cross-region certificate, HostedZone is a required attribute. For our purposes, we have our HostedZone in another AWS account. This seems like an odd limitation as I can create a certificate in another region through the AWS console.

**Comment by salihoglu87** [comment]
The "acm.DnsValidatedCertificate" is now deprecated. Is there any movement so we can create cross-region certificate using "acm.Certificate"? 
What is the difference between "acm.DnsValidatedCertificate" and "acm.Certificate" so that we can't set region by "acm.Certificate"?

**Comment by peterfranzen** [comment]
I'm also struggling with this.  I'm trying to deploy a S3/CloudFront stack in us-west-2 with a certificate that was created in another stack in us-east-1.  Short of writing a custom resource lambda to go hunt for the certificate and put its info into SSM Parameter Store, what is the best way to accomplish this?

**Comment by mostafafarzaneh** [comment]
The problem becomes a real nightmare when using a custom domain for Cognito UserPool. The UserPool CustomDomain needs an ICertificate, not a Certificate ARN. So there is no solution besides DnsValidatedCertificate which is deprecated.
The CloudFront accepts the Certificate ARN, which you can somehow work around with a custom resource. But not in the Cognito Custom domain case.

**Comment by LukaHedt** [comment]
Yeah this is going to become a _real_ problem using certs for Cloudfront too.
Requiring downtime for playing with cloudformation stack regions (because of a construct deprecation) is not very fun DevOps.
Also makes for some very, _very_ annoying CDK infrastructure dependencies, which drives me insane.

**Comment by Brian-Azizi** [comment]
Why has `DnsValidatedCertificate` been deprecated? How is it's usecase solved with `Certificate` (which is the suggested alternative)?

**Comment by vincent-milia** [comment]
> Why has `DnsValidatedCertificate` been deprecated? How is it's usecase solved with `Certificate` (which is the suggested alternative)?

y I just came across the same problem. I believe one option to solve this is to create the certificate in a separate stack (with env.region: us-east-1). Then reference the certificate in the other stack and use lookup. Don't know if there is a better way.

**Comment by mostafafarzaneh** [comment]
> > Why has `DnsValidatedCertificate` been deprecated? How is it's usecase solved with `Certificate` (which is the suggested alternative)?
> 
> y I just came across the same problem. I believe one option to solve this is to create the certificate in a separate stack (with env.region: us-east-1). Then reference the certificate in the other stack and use lookup. Don't know if there is a better way.

Is there a function for lookup in Certificate construct? there is `fromCertificateArn`, but you cannot use it with cross-region certificates.

**Comment by sensedata1** [comment]
We now also have a lot of refactoring to do as it appears that the deprecated construct will no longer create the resources defined in our stacks. If one of the certificates gets deleted, we're currently unable to just re-deploy to replace it. Not best pleased.

**Comment by mhkafadar** [comment]
Hello,

I've successfully implemented a cross-region deployment, positioning the S3-Cloudfront stack in Europe and the certificate in the us-east-1 region. I've detailed my approach in a blog post. https://medium.com/@mhkafadar/a-practical-aws-cdk-walkthrough-deploying-multiple-websites-to-s3-and-cloudfront-7caaabc9c327

**Comment by epiphone** [comment]
> > Why has `DnsValidatedCertificate` been deprecated? How is it's usecase solved with `Certificate` (which is the suggested alternative)?
> 
> y I just came across the same problem. I believe one option to solve this is to create the certificate in a separate stack (with env.region: us-east-1). Then reference the certificate in the other stack and use lookup. Don't know if there is a better way.

This is now documented in https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_certificatemanager-readme.html#cross-region-certificates.

**Comment by DanielLaberge** [comment]
This is quite an annoying pitfall that everyone gets into if using CloudFront with HTTPS and not deploying resources in us-east-1.
Shouldn't solving this kind of headache be CDK's raison d'être?

**Comment by rredpoppy** [comment]
this is just downrigth awful..

**Comment by case303** [comment]
Thanks @JavierMendozaGomez  and @epiphone 
This provided good insight around: scope, stacks, and regions.
Specify the desired region in a new Stack, then pass that stack into the resource's scope. As best illustrated in the docs:  [Cross-region Certificates](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_certificatemanager-readme.html#cross-region-certificates.)
```
const certificateStack = new cdk.Stack(this, CERTIFICATE_STACK_ID,
      {
        env: {
          account: process.env.AWS_ACCOUNT_ID || process.env.CDK_DEFAULT_ACCOUNT,
          region: 'us-east-1'
        },
        crossRegionReferences: true,
      });
      
const certificate = new acm.Certificate(certificateStack, CERTIFICATE_ID, {
      certificateName: CERTIFICATE_ID,
      domainName: DOMAIN_NAME,
      subjectAlternativeNames: [`*.${DOMAIN_NAME}`],
      validation: acm.CertificateValidation.fromDns(hostedZone),
    });
```

**Comment by pqnet** [comment]
Unfortunately having a cross-stack reference has the limitation of having to hard-code the zone for all other stacks in the app because `Cross stack/region references are only supported for stacks with an explicit region defined.`. It would be great if any of this would be true:
- we could have the certificate in `us-east-1` while its stack is deployed somewhere else (as it is for `DnsValidatedCertificate`). Why has this possibility been dropped?
- CloudFront didn't require us to host the certificate in `us-east-1`.
- cross-stack references wouldn't require to explicitly specify the stack at synth time

**Comment by quantfreedom** [comment]
I have finally figured out how to actually do this ... i am using cdk 2.128.0

This actually works because i am in ca-central-1 and i was able to do it in one stack and created a us-east-1 cert

you can checkout the github code here https://github.com/quantfreedom/aws_cdk_testing/blob/main/amazon_cdk/frontend/www.py

if that doesn't work here is a gist of the current working version https://gist.github.com/quantfreedom/e71267553edc8e0760e88d48ad8b45a7

I watched this youtube video to get an understanding of what is going on and why i am doing the things i am doing 
https://www.youtube.com/watch?v=p6Os-_t0gEs

and here is their gh code https://github.com/Durgaprasad-Budhwani/hands-on-aws-cdk-lab/blob/main/ts/cloudfront/lib/cloudfront-stack.ts

**Comment by lucashfreitas** [comment]
This is an example of a really bad experience from with AWS certificate manager.

Got over this issue a number of times which might have costed us AWS developers hundred of hours.

Ended up creating the certificate manually via CLI/console and use the ARN to reference it.

**Comment by pahud** [comment]
Hi, 

I am reopening this issue as we've seen the concerns are still outstanding. Let me clarify this and we welcome any further thoughts to make the CDK experience even better.

## Major complaints about current experience

1. Cross-Region Certificate Management Difficulties:
- Struggle with creating certificates in us-east-1 while using them in different regions
- Particularly problematic for services like CloudFront and Cognito that require certificates to be in us-east-1
- Cross-stack references don't work well across regions

2. Deprecation of DnsValidatedCertificate:
- The DnsValidatedCertificate construct was deprecated without a clear alternative solution
- Being frustrated that the replacement (Certificate construct) doesn't handle cross-region cases as well
- This has forced teams to refactor their existing infrastructure code

3. Limited Solutions:
- No straightforward "CDK way" to handle cross-region certificates
- Have to resort to workarounds like:
  - Creating separate stacks
  - Using SSM Parameter Store
  - Manual certificate creation
  - Custom resources

4. Specific Service Integration Issues:
- Cognito UserPool CustomDomain requires an ICertificate, not just a Certificate ARN, making workarounds more difficult
- CloudFront deployments requiring certificates from us-east-1 create unnecessary complexity

5. Developer Experience:
- Many users express that this should be a core functionality of CDK
- The current situation forces developers to spend significant time finding workarounds


Thank you for raising this important concern about cross-region certificate management. We understand this has been a pain point for many of our customers, and we appreciate the detailed feedback.

## Current Recommended Solution

For cross-region certificate usage (such as with CloudFront and Cognito), we currently recommend using a multi-stack deployment pattern with the `crossRegionReferences` feature. Here's the recommended approach:

```ts
const stack1 = new Stack(app, 'CertStack', {
  env: { region: 'us-east-1' }, // Certificate must be in us-east-1 for CloudFront
  crossRegionReferences: true
});

const cert = new acm.Certificate(stack1, 'Cert', {
  domainName: '*.example.com',
  validation: acm.CertificateValidation.fromDns() 
});

const stack2 = new Stack(app, 'DistributionStack', {
  env: { region: 'us-east-2' }, // Your app can be in any region
  crossRegionReferences: true
});

// Reference the certificate from stack1 in stack2
new cloudfront.Distribution(stack2, 'Distribution', {
  defaultBehavior: { origin: new origins.HttpOrigin('example.com') },
  domainNames: ['dev.example.com'],
  certificate: cert
});
```

The reason you need to specify the `region` code in `env` is because this demo code is deploying two stacks in different regions in a shared CDK app.  You can separate them into two different CDK apps for different region and use `process.env.CDK_DEFAULT_REGION` for the region env so you don't have to explicitly specify the region code.


## Important Notes

1. While this feature is marked as experimental, it represents our current best practice for handling cross-region certificates
2. We acknowledge that the deprecation of `DnsValidatedCertificate` has caused friction, and we're working to improve the experience
3. The `crossRegionReferences` feature aims to provide a more maintainable solution compared to previous workarounds

## Moving Forward

We understand that the current solution might not cover all use cases perfectly, and we're actively working on improvements. Your feedback helps us prioritize enhancements to the CDK's certificate management capabilities.

If you encounter specific issues with this approach or have additional use cases not covered by the current solution, please let us know. We're committed to improving the developer experience and making cross-region certificate management more straightforward.

## Additional Resources

- [Official CDK Certificate Manager documentation](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_certificatemanager-readme.html#cross-region-certificates)
- [AWS CDK API Reference](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_certificatemanager.Certificate.html)

Also, as this thread has been very lengthy and could be very difficult to help users focus on the latest solution. If you are still having any issue or feature request around this use case, please kindly create a new issue. We will be happy to look into your case or any new idea about the API design.

Please don't hesitate to provide additional feedback or ask for clarification. We're here to help!

**Comment by jk2l** [comment]
@pahud do this cross stack reference work if there requirement of replacement of ACM?

e.g. domain name expanded and added a new subdomain. using `DnsValidatedCertificate` will allow us follow Cfn flow (Update with replacement, update dependency, clean up old resource)

i don't see how cross stack reference can handle this well

**Comment by pahud** [comment]
@jk2l 

Thank you for sharing your use case. I think before CFN has native cross-region support, any workaround in CDK would never have the best developer experience. I see 120 upvotes on https://github.com/aws-cloudformation/cloudformation-coverage-roadmap/issues/523 and I believe CFN team definitely would consider that but I can't see any further details from there. Yes, I'd submit an internal ticket to CFN team about this for better visibility but I can't guarantee any timelines about its ETA of that CFN support. I'll add internal ticket reference here and I'll update here as soon as I have any news I can share.

**Comment by pahud** [comment]
internal tracking - V1660274068

**Comment by rantoniuk** [comment]
Until a satisfying (for the community = actual clients) solution is developed, it would be great to be able to silence the:

```
[WARNING] aws-cdk-lib.aws_certificatemanager.DnsValidatedCertificate is deprecated.
  use {@link Certificate } instead
  This API will be removed in the next major release.
```

via [cdk acknowledge](https://docs.aws.amazon.com/cdk/v2/guide/ref-cli-cmd-ack.html).

@pahud @peterwoodworth, happy to provide a PR if there is a consensus that the deprecation decision is going to be re-visited .

**Comment by jk2l** [comment]
> Until a satisfying (for the community = actual clients) solution is developed, it would be great to be able to silence the:
> 
> ```
> [WARNING] aws-cdk-lib.aws_certificatemanager.DnsValidatedCertificate is deprecated.
>   use {@link Certificate } instead
>   This API will be removed in the next major release.
> ```
> 
> via [cdk acknowledge](https://docs.aws.amazon.com/cdk/v2/guide/ref-cli-cmd-ack.html).
> 
> [@pahud](https://github.com/pahud) [@peterwoodworth](https://github.com/peterwoodworth), happy to provide a PR if there is a consensus that the deprecation decision is going to be re-visited .

+1

**Comment by rantoniuk** [comment]
@njlynch @pahud care to provide an update from internal ticket V1660274068 on what's your plan (or what should be the plan for the community)? 
I do agree that all the semi-solutions and workarounds provided are over-engineered and caused by a design fault on AWS side (or an initial rush to PROD with a new service and never improve UX and availability from the initial east-1 to other regio)ns).

**Comment by pahud** [comment]
Hi @rantoniuk and @AnharHussainMiah,

I've escalated this issue internally and am actively working to connect with the feature owners. I'll provide an update here if there are any details we can share publicly at this time.

To ensure this is tracked effectively, I've assigned the issue to myself and will follow up with the relevant internal teams.

**Comment by pahud** [comment]
bumped to p1

**Comment by Mati365** [comment]
@pahud what P1 mean?

**Comment by pahud** [comment]
@Mati365 

`p1` is used for issues or PRs from the team, indicating high-priority items that require immediate attention.

In the [CDK Contributing Guide](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md), all issues with 20 upvotes will be automatically bumped to `p1` for team's visibility. This issue is having 21 upvotes but for some reason our workflow didn't successfully bump it so I just manually made it a `p1`.

While this issue is still pending on other internal teams before CDK can bring this into reality, we are very actively monitoring this issue. I'll circle back here when I get any updates.

**Comment by pahud** [comment]
HI @blimmer 

Let me summarize this issue and propose a solution here.

# Cross-Region Certificate Management in AWS CDK

## 1. Pain Points

- **Region-Specific Requirements**: Services like CloudFront and Cognito require certificates to be in `us-east-1`, even when the rest of the infrastructure is in a different region
- **Cross-Region Reference Limitations**: CDK doesn't easily support referencing resources across regions
- **Complex Workarounds**: Current solutions require:
  - Multiple stacks with explicit region configurations
  - Using the experimental `crossRegionReferences: true` feature
  - Custom resources or SSM parameters to pass certificate ARNs
- **Deprecated Solutions**: The `DnsValidatedCertificate` construct that previously helped with this use case has been deprecated
- **Developer Experience Impact**: Significant time spent finding workarounds for what should be a straightforward use case

## 2. Expected User Experience

Developers expect a simple, intuitive way to:

- Define a certificate in `us-east-1` once
- Reference it from resources in any region without special configuration
- Avoid splitting infrastructure into multiple stacks just for region-specific requirements
- Have CDK handle the cross-region complexity internally
- Use a consistent pattern across all AWS services with region-specific requirements
- Not have to deal with breaking changes when solutions are deprecated

## 3. Proposed Solution: `Certificate.fromGlobalLookup()`

A new method on the `Certificate` class that would:

```typescript
// Simple API that always looks up certificates in us-east-1
const certificate = acm.Certificate.fromGlobalLookup(this, 'MyCert', {
  domainName: 'example.com',
});

// Use it directly in Cognito in any region
new cognito.UserPoolDomain(this, 'Domain', {
  userPool,
  customDomain: {
    domainName: 'auth.example.com',
    certificate
  }
});
```

### How It Works
- During synthesis/deployment, the CDK makes an AWS SDK call to us-east-1
- It queries ACM for certificates matching the provided domain name
- It returns an ICertificate implementation containing the ARN
- This implementation is usable by any service requiring a certificate, regardless of region

### Benefits
- Single Stack Solution: No need for separate stacks in different regions
- No Cross-Region References: Avoids the experimental crossRegionReferences: true feature
- Simplified Developer Experience: Mirrors familiar CDK lookup patterns
- Synthesis-Time Resolution: Certificate ARN resolved during synthesis, not deployment
- Caching: Results cached in cdk.context.json for future deployments

### Implementation Flow
1. Deploy a stack in us-east-1 that creates and validates the certificate
2. Run cdk synth for the stack in another region (performs the lookup)
3. Deploy the second stack which references the certificate

This approach maintains the same deployment sequence but eliminates the need for cross-region references in CloudFormation, simplifying the architecture and avoiding experimental features.


I am drawing a mermaid to explain this flow

```mermaid
flowchart TD
    subgraph "Step 1: Certificate Creation"
        A["Deploy Stack in us-east-1"] --> B["Create ACM Certificate"]
        B --> C["Validate Certificate via DNS/Email/HTTP"]
        C --> D["Certificate is ISSUED"]
    end
    
    subgraph "Step 2: Certificate Lookup"
        E["Run cdk synth for Stack in eu-central-1"] --> F["fromGlobalLookup() executes"]
        F --> G["AWS SDK call to us-east-1"]
        G --> H["Query ACM for matching certificate"]
        H --> I["Certificate ARN found"]
        I --> J["ARN stored in cdk.context.json"]
    end
    
    subgraph "Step 3: Certificate Usage"
        K["Deploy Stack in eu-central-1"] --> L["CloudFormation creates resources"]
        L --> M["Resources reference Certificate ARN"]
        M --> N["Cognito/CloudFront uses certificate from us-east-1"]
    end
    
    D --> F
    J --> K
```

## Advantages of the `Certificate.fromGlobalLookup()` Solution

### 1. Simplified Developer Experience

- **Intuitive API**: Clean, simple method that clearly communicates its purpose
- **Reduced Cognitive Load**: Developers don't need to understand cross-region reference complexities
- **Familiar Pattern**: Follows established CDK lookup patterns (similar to `Vpc.fromLookup()`)
- **Self-Documenting**: Method name itself documents the solution to a common AWS architectural pattern

### 2. Technical Advantages

- **No Experimental Features**: Avoids the experimental `crossRegionReferences: true` feature flag
- **No CloudFormation Cross-Stack References**: Certificate ARN is resolved at synthesis time, not deployment time
- **Independent Stack Deployments**: Stacks can be deployed independently once the certificate exists
- **Caching**: Results stored in `cdk.context.json`, improving subsequent deployment speed
- **Synthesis-Time Resolution**: Issues are detected early in the development cycle, not during deployment

### 3. Architectural Improvements

- **Single Stack Deployment Option**: No need to split infrastructure into multiple stacks just for certificates
- **Cleaner CloudFormation Templates**: Generated templates don't contain complex cross-region references
- **Reduced Deployment Complexity**: Fewer stacks to manage and deploy
- **Better Error Handling**: Clear error messages when certificates don't exist or multiple matches are found

### 4. Operational Benefits

- **Consistent with AWS Best Practices**: Aligns with AWS's recommendation to create certificates in us-east-1 for global services
- **Backward Compatible**: Works with existing certificates already created in us-east-1
- **Forward Compatible**: Will continue to work even as AWS evolves their cross-region capabilities
- **No Breaking Changes**: Unlike the deprecation of `DnsValidatedCertificate`, this approach adds functionality without removing it

### 5. Implementation Advantages

- **Minimal Code**: Can be implemented with relatively few lines of code
- **Reusable Pattern**: The same approach could be extended to other cross-region resource types
- **Testable**: Easier to mock and test than complex cross-region CloudFormation references
- **Maintainable**: Follows established CDK patterns, making it easier for the community to understand and maintain



Wondering if this is something we expect?

**Comment by jk2l** [comment]
> HI [@blimmer](https://github.com/blimmer)
> 
> Let me summarize this issue and propose a solution here.
> # Cross-Region Certificate Management in AWS CDK
> ## 1. Pain Points
> 
>     * **Region-Specific Requirements**: Services like CloudFront and Cognito require certificates to be in `us-east-1`, even when the rest of the infrastructure is in a different region
> 
>     * **Cross-Region Reference Limitations**: CDK doesn't easily support referencing resources across regions
> 
>     * **Complex Workarounds**: Current solutions require:
>       
>       * Multiple stacks with explicit region configurations
>       * Using the experimental `crossRegionReferences: true` feature
>       * Custom resources or SSM parameters to pass certificate ARNs
> 
>     * **Deprecated Solutions**: The `DnsValidatedCertificate` construct that previously helped with this use case has been deprecated
> 
>     * **Developer Experience Impact**: Significant time spent finding workarounds for what should be a straightforward use case
> 
> 
> ## 2. Expected User Experience
> 
> Developers expect a simple, intuitive way to:
> 
>     * Define a certificate in `us-east-1` once
> 
>     * Reference it from resources in any region without special configuration
> 
>     * Avoid splitting infrastructure into multiple stacks just for region-specific requirements
> 
>     * Have CDK handle the cross-region complexity internally
> 
>     * Use a consistent pattern across all AWS services with region-specific requirements
> 
>     * Not have to deal with breaking changes when solutions are deprecated
> 
> 
> ## 3. Proposed Solution: `Certificate.fromGlobalLookup()`
> 
> A new method on the `Certificate` class that would:
> 
> // Simple API that always looks up certificates in us-east-1
> const certificate = acm.Certificate.fromGlobalLookup(this, 'MyCert', {
>   domainName: 'example.com',
> });
> 
> // Use it directly in Cognito in any region
> new cognito.UserPoolDomain(this, 'Domain', {
>   userPool,
>   customDomain: {
>     domainName: 'auth.example.com',
>     certificate
>   }
> });
> 
> ### How It Works
> 
>     * During synthesis/deployment, the CDK makes an AWS SDK call to us-east-1
> 
>     * It queries ACM for certificates matching the provided domain name
> 
>     * It returns an ICertificate implementation containing the ARN
> 
>     * This implementation is usable by any service requiring a certificate, regardless of region
> 
> 
> ### Benefits
> 
>     * Single Stack Solution: No need for separate stacks in different regions
> 
>     * No Cross-Region References: Avoids the experimental crossRegionReferences: true feature
> 
>     * Simplified Developer Experience: Mirrors familiar CDK lookup patterns
> 
>     * Synthesis-Time Resolution: Certificate ARN resolved during synthesis, not deployment
> 
>     * Caching: Results cached in cdk.context.json for future deployments
> 
> 
> ### Implementation Flow
> 
>     1. Deploy a stack in us-east-1 that creates and validates the certificate
> 
>     2. Run cdk synth for the stack in another region (performs the lookup)
> 
>     3. Deploy the second stack which references the certificate
> 
> 
> This approach maintains the same deployment sequence but eliminates the need for cross-region references in CloudFormation, simplifying the architecture and avoiding experimental features.
> 
> I am drawing a mermaid to explain this flow
> ## Advantages of the `Certificate.fromGlobalLookup()` Solution
> ### 1. Simplified Developer Experience
> 
>     * **Intuitive API**: Clean, simple method that clearly communicates its purpose
> 
>     * **Reduced Cognitive Load**: Developers don't need to understand cross-region reference complexities
> 
>     * **Familiar Pattern**: Follows established CDK lookup patterns (similar to `Vpc.fromLookup()`)
> 
>     * **Self-Documenting**: Method name itself documents the solution to a common AWS architectural pattern
> 
> 
> ### 2. Technical Advantages
> 
>     * **No Experimental Features**: Avoids the experimental `crossRegionReferences: true` feature flag
> 
>     * **No CloudFormation Cross-Stack References**: Certificate ARN is resolved at synthesis time, not deployment time
> 
>     * **Independent Stack Deployments**: Stacks can be deployed independently once the certificate exists
> 
>     * **Caching**: Results stored in `cdk.context.json`, improving subsequent deployment speed
> 
>     * **Synthesis-Time Resolution**: Issues are detected early in the development cycle, not during deployment
> 
> 
> ### 3. Architectural Improvements
> 
>     * **Single Stack Deployment Option**: No need to split infrastructure into multiple stacks just for certificates
> 
>     * **Cleaner CloudFormation Templates**: Generated templates don't contain complex cross-region references
> 
>     * **Reduced Deployment Complexity**: Fewer stacks to manage and deploy
> 
>     * **Better Error Handling**: Clear error messages when certificates don't exist or multiple matches are found
> 
> 
> ### 4. Operational Benefits
> 
>     * **Consistent with AWS Best Practices**: Aligns with AWS's recommendation to create certificates in us-east-1 for global services
> 
>     * **Backward Compatible**: Works with existing certificates already created in us-east-1
> 
>     * **Forward Compatible**: Will continue to work even as AWS evolves their cross-region capabilities
> 
>     * **No Breaking Changes**: Unlike the deprecation of `DnsValidatedCertificate`, this approach adds functionality without removing it
> 
> 
> ### 5. Implementation Advantages
> 
>     * **Minimal Code**: Can be implemented with relatively few lines of code
> 
>     * **Reusable Pattern**: The same approach could be extended to other cross-region resource types
> 
>     * **Testable**: Easier to mock and test than complex cross-region CloudFormation references
> 
>     * **Maintainable**: Follows established CDK patterns, making it easier for the community to understand and maintain
> 
> 
> Wondering if this is something we expect?

Ideally i would prefer not to deploy stack in another region. In my use case my client mainly are government related. and they have very specific least privilege permission. By default only the government operation country is enabled in AWS Control Tower and other region are restricted for compliance purpose.

Atm we don't need to open cloudformation in us-east-1, we only need to open ACM. with your solution here, i will need to enable CloudFormation for us-east-1 too. While in theory we can manually provision the certificate in us-east-1 but that is also mean introduce human errors plus prod environment often restricted for direct access outside of Infrastructure as Code.

**Comment by dbartholomae** [comment]
Hi!

That’s unfortunately missing a major pain point:
Resources need to be transparentes encapsulated.

With the currently deprecated DnsValidatedCertificate, it is possible to create reusable constructs like a `StaticWebsite` that takes care of creating an S3 bucket, aCloudFront distribution, an S3 deployment, a certificate and a DNS record, encapsulating the full website. This wouldn’t be possible with a global lookup.

**Comment by rantoniuk** [comment]
> I wonder if [Automated HTTP validated public certificates with Amazon CloudFront](https://aws.amazon.com/about-aws/whats-new/2025/04/automated-http-validated-public-certificates-amazon-cloudfront/), just announced a few days ago, could help with the cross-region CloudFront certificate issue?

_Since the HTTP validation token works for any AWS Region where CloudFront is available, you can re-create the same certificate in multiple Regions._ 

@pahud If the above sentence means what I understand it means, why are we still talking about keeping certificates in us-east-1 ? As mentioned above, the regions should be autonomous and for compliancy reasons I wouldn't want to enable any _special_ services in us-east-1. 
Apart from compliancy it makes it so much harder in terms of FinOps, code maintenance etc.

**Comment by blimmer** [comment]
I agree with everything folks have said here. From my perspective, the deprecated `DnsValidatedCertificate` _is_ exactly what I want. I don't have to think about multiple stacks/global lookups/etc.

From a DX perspective, I just `new DnsValidatedCertificate(this, 'Cert', { region: 'us-east-1' });` and it works perfectly. All of these other options put an additional burden on the developer. I understand that the Custom Resource has been a pain for the CDK team and has some strange corner cases, but I've never had a problem with it.

If `DnsValidatedCertificate` goes away, the replacement should be as easy as what we could do before (if not easier).

This is why I've kept pushing the CloudFormation-native fix: https://github.com/aws-cloudformation/cloudformation-coverage-roadmap/issues/523, but it's unclear to me if the CloudFormation team would do that or not.

**Comment by burakalakus** [comment]
@ahmetozergetir also reported another use case (#36441) with same limitation. 

**Scenario:** Using CloudFront VPC Origins to expose internal Application Load Balancers (ALBs) that are provisioned in different regions
**Architecture:** Internal ALB managed by EKS ALB Controller in an EU region, with CloudFront distribution in us-east-1 needing to reference it via VPC Origin
**Current Limitation:** Cannot use `ApplicationLoadBalancer.fromLookup()` for cross-region resources because CDK's context provider only looks up resources in the stack's configured region during synthesis time

CloudFront (global service) must be deployed in us-east-1, but workloads and ALBs often exist in other regions.
This is a common pattern for exposing internal services globally while maintaining regional deployments.
Current workaround requires manually creating VPC Origins and using `fromVpcOriginArn()`, which bypasses CDK's type safety and automation benefits.
